compiler/rustc_lint/src/context.rs

```diff
@@ -768,22 +768,56 @@ pub trait LintContext: Sized {
-                    let possibilities: Vec<Symbol> = names_valid.iter().map(|s| *s).collect();
+                BuiltinLintDiagnostics::UnexpectedCfgName((name, name_span), value) => {
+                    let possibilities: Vec<Symbol> = sess.parse_sess.check_config.expecteds.keys().map(|s| *s).collect();
 
                     // Suggest the most probable if we found one
                     if let Some(best_match) = find_best_match_for_name(&possibilities, name, None) {
-                        db.span_suggestion(name_span, "did you mean", best_match, Applicability::MaybeIncorrect);
+                        if let Some(rustc_session::config::ExpectedValues::Some(best_match_values)) =
+                            sess.parse_sess.check_config.expecteds.get(&best_match) {
+                            let mut possibilities = best_match_values.iter()
+                                .flatten()
+                                .map(Symbol::as_str)
+                                .collect::<Vec<_>>();
+                            possibilities.sort();
+
+                            if let Some((value, value_span)) = value {
+                                if best_match_values.contains(&Some(value)) {
+                                    db.span_suggestion(name_span, "there is an config with a similar name and value", best_match, Applicability::MaybeIncorrect);
+                                } else if best_match_values.contains(&None) {
+                                    db.span_suggestion(name_span.to(value_span), "there is an config with a similar name and no value", best_match, Applicability::MaybeIncorrect);
+                                } else {
+                                    if let Some(first_value) = possibilities.first() {
+                                        db.span_suggestion(name_span.to(value_span), "there is an config with a similar name and different values", format!("{best_match} = \"{first_value}\""), Applicability::MaybeIncorrect);
+                                    } else {
+                                        db.span_suggestion(name_span.to(value_span), "there is an config with a similar name and different values", best_match, Applicability::MaybeIncorrect);
+                                    };
+                                }
+                            } else {
+                                db.span_suggestion(name_span, "there is an config with a similar name", best_match, Applicability::MaybeIncorrect);
+                            }
+
+                            if !possibilities.is_empty() {
+                                let possibilities = possibilities.join("`, `");
+                                db.help(format!("expected values for `{best_match}` are: `{possibilities}`"));
+                            }
+                        } else {
+                            // being here should not be possible, but better show something a bit
+                            // wrong than an ICE
```

Comment:
If it isn't possible then it should indeed be an ICE, otherwise it needs a test case.

Reply:
After some testing it is actually possible to hit this case and is already covered by the `well-known-names.rs` test.